### PR TITLE
support workflow, workflow imports, workflow steps, inputs and outputs

### PR DESCRIPTION
**Pull request by Th3nn3ss**

### CODE REVIEW

File: wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+      --loglevel=DEBUG
+
+    sort -b -r -n -k3 "${reads_basename}.count.${ref_basename}.txt.unsorted" > "${reads_basename}.count.${ref_basename}.txt"
+    head -n ${topNHits} "${reads_basename}.count.${ref_basename}.txt" > "${reads_basename}.count.${ref_basename}.top_${topNHits}_hits.txt"
+    head -1 "${reads_basename}.count.${ref_basename}.txt" | cut -f 1 > "${reads_basename}.count.${ref_basename}.top.txt"
+  }
+
+  output {
+    File   report           = "${reads_basename}.txt"
+    # File   report           = "${reads_basename}.count.${ref_basename}.txt"
```
Comment (mr-c) [inline wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl:307]:
```suggestion
    File   report           = "${reads_basename}.txt"
    # File   report           = "${reads_basename}.count.${ref_basename}.txt"
```
FYI: This is needed for the tool to work

File: wdl2cwl/tests/cwl_files/align_and_count.cwl [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.txt')
+          - id: report_top_hits
+            type: File
+            outputBinding:
+                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.txt')
+          - id: top_hit_id
+            type: string
+            outputBinding:
+                loadContents: true
+                glob: nputs.reads_bam.basename.replace(/\.bam$/, '')  + '.txt
```
Comment (mr-c) [inline wdl2cwl/tests/cwl_files/align_and_count.cwl:83]:
```suggestion
                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.txt
```

**Review by mr-c — CHANGES_REQUESTED** [review]

File: wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl [hunk 3]
```diff
@@ -0,0 +1,538 @@
+version 1.0
+
+# Source: https://github.com/broadinstitute/viral-pipelines/blob/master/pipes/WDL/tasks/tasks_reports.wdl
```
Comment (mr-c) [inline wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl:3]:
```suggestion
# Source: https://github.com/broadinstitute/viral-pipelines/blob/2a24ae82bf3c39dbbbfb8efdac0f7d276e780a4e/pipes/WDL/tasks/tasks_reports.wdl
```
To get a permalink, on almost any GitHub page hit the "y" key and then copy the updated location

File: wdl2cwl/tests/wdl_files/align_and_count.wdl [hunk 4]
```diff
@@ -0,0 +1,21 @@
+version 1.0
+
+# Source: https://github.com/broadinstitute/viral-pipelines/blob/master/pipes/WDL/workflows/align_and_count.wdl
+
+import "/home/dennis/Documents/CWLab/wdl-cwl-translator/wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl" as reports
+# import "../tasks/tasks_reports.wdl" as reports
```
Comment (mr-c) [inline wdl2cwl/tests/wdl_files/align_and_count.wdl:6]:
```suggestion
import "./tasks/tasks_reports.wdl" as reports
```
In unix / shell `..` means the parent directory, but the `tasks` folder is in THIS folder (`.`)

File: wdl2cwl/main.py [hunk 5]
```diff
@@ -60,10 +61,66 @@ def load_wdl_objects(
             return self.load_wdl_task(obj)
         raise Exception(f"Unimplemented type: {type(obj)}: {obj}")
 
+    def load_wdl_workflow(self, obj: WDL.Tree.Workflow) -> cwl.Workflow:
+        """Load WDL workflow and convert to CWL."""
+        inputs: list[cwl.WorkflowInputParameter] = []
+        outputs: list[cwl.WorkflowOutputParameter] = []
+        wf_steps: list[cwl.WorkflowStep] = []
+        wf_name = obj.name
```
Comment (mr-c) [inline wdl2cwl/main.py:69]:
```suggestion
        inputs: List[cwl.WorkflowInputParameter] = []
        outputs: List[cwl.WorkflowOutputParameter] = []
        wf_steps: List[cwl.WorkflowStep] = []
        wf_name = obj.name
```

File: wdl2cwl/tests/cwl_files/align_and_count.cwl [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            outputBinding:
+                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.count.'
+                    + inputs.ref_db.basename.replace(/\.fasta$/, '')  + '.top_' +
+                    inputs.topNHits + '_hits.txt')
+          - id: top_hit_id
+            type: string
+            outputBinding:
+                loadContents: true
+                glob: nputs.reads_bam.basename.replace(/\.bam$/, '')  + '.count.'
+                    + inputs.ref_db.basename.replace(/\.fasta$/, '')  + '.top.txt
```
Comment (mr-c) [inline wdl2cwl/tests/cwl_files/align_and_count.cwl:87]:
```suggestion
                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.count.'
                    + inputs.ref_db.basename.replace(/\.fasta$/, '')  + '.top.txt')
```
odd truncation here...

File: wdl2cwl/tests/cwl_files/align_and_count.cwl [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.count.'
+                    + inputs.ref_db.basename.replace(/\.fasta$/, '')  + '.top_' +
+                    inputs.topNHits + '_hits.txt')
+          - id: top_hit_id
+            type: string
+            outputBinding:
+                loadContents: true
+                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.count.'
+                    + inputs.ref_db.basename.replace(/\.fasta$/, '')  + '.top_' +
+                    inputs.topNHits + '_hits.txt')
```
Comment (mr-c) [inline wdl2cwl/tests/cwl_files/align_and_count.cwl:88]: Odd, this is the correct glob for `report_top_hits`, but not `top_hit_id`
```suggestion
                glob: $(inputs.reads_bam.basename.replace(/\.bam$/, '')  + '.count.'
                    + inputs.ref_db.basename.replace(/\.fasta$/, '')  + '.top.txt')
```

File: wdl2cwl/main.py [hunk 8]
```diff
@@ -60,10 +61,66 @@ def load_wdl_objects(
...
+                for x in cwl_call_inputs
+            ]
+            inputs.extend(
+                [
+                    cwl.WorkflowInputParameter(
+                        id=f"{namespace}.{call_name}.{x.id}",
+                        type=x.type,
+                        default=x.default,
+                    )
+                    for x in cwl_call_inputs
```
Comment (kinow) [inline wdl2cwl/main.py:87]: Ah! I think this is an excellent case for when not to use list comprehensions :nerd_face: 

You are iterating `cwl_call_inputs`, and `cwl_call_ouputs` twice to build the outputs and parameters. Instead, with a `for` loop you can easily iterate it once and populate both lists :+1: 

I believe we could even do that in one go with list comprehensions too, but that wouldn't be easy to read & maintain I think.

**Comment by mr-c** [comment]
Okay, the new test failed because it needs ~15GiB memory 😛 

In https://github.com/common-workflow-lab/wdl-cwl-translator/pull/138/commits/e5eadc01b363b1c8e4c5be2ed92051c8b2d052a9 I moved the test out of the automatic testing, but I can confirm locally that it succeeds.

Hopefully we can find smaller test data or another workflow that is testable under GitHub Actions in a future PR.

**Review by mr-c — APPROVED** [review]

File: wdl2cwl/main.py [hunk 9]
```diff
@@ -275,7 +332,8 @@ def get_literal_name(
         # the literal value is what's needed
         if isinstance(expr.parent, WDL.Expr.Apply):  # type: ignore
             return expr.literal.value  # type: ignore
-        raise Exception(f"The parent expression for {expr} is not WDL.Expr.Apply")
+        parent_name = expr.parent.name  # type: ignore
+        return self.get_input(parent_name)
```
Comment (mr-c) [inline wdl2cwl/main.py:336]: These two lines have no test coverage. Maybe you should save them for another pull request?

File: wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+    File   report           = "${reads_basename}.count.${ref_basename}.txt"
+    File   report_top_hits  = "${reads_basename}.count.${ref_basename}.top_${topNHits}_hits.txt"
+    String top_hit_id       = read_string("${reads_basename}.count.${ref_basename}.top.txt")
+    String viralngs_version = read_string("VERSION")
+  }
+
+  runtime {
+    memory: "15 GB"
+    cpu: 4
+    docker: docker
```
Comment (mr-c) [inline wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl:315]: This doesn't make sense, `docker` is not the name of a valid container

Shouldn't it be what was in the original file?
```suggestion
    docker: "~{docker}"
```

File: wdl2cwl/main.py [hunk 11]
```diff
@@ -60,10 +61,66 @@ def load_wdl_objects(
             return self.load_wdl_task(obj)
         raise Exception(f"Unimplemented type: {type(obj)}: {obj}")
 
+    def load_wdl_workflow(self, obj: WDL.Tree.Workflow) -> cwl.Workflow:
+        """Load WDL workflow and convert to CWL."""
+        inputs: List[cwl.WorkflowInputParameter] = []
```
Comment (mr-c) [inline wdl2cwl/main.py:66]: For getting the workflow inputs, have you considered looking at `obj.available_inputs`? 
https://miniwdl.readthedocs.io/en/latest/WDL.html#WDL.Tree.Workflow.available_inputs

That will probably work better for workflows with many layers of nesting and imports as MiniWDL will figure out all of that for you
``` python
for wf_input in obj.available_inputs:
  info: Any = wf_input.info  # Always None?
  name: str = wf_input.name   # like `align_and_count.reads_bam`
  value: WDL.Tree.Decl = wf_input.value  # like WDL.Tree.Decl(type=WDL.Type.File, name="reads_bam")
```

File: wdl2cwl/main.py (hunk 9 above)
Comment (Th3nn3ss) [reply wdl2cwl/main.py:336]: It's covered in the new file. It's needed for top_hit_id

File: wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl (hunk 10 above)
Comment (Th3nn3ss) [reply wdl2cwl/tests/wdl_files/tasks/tasks_reports.wdl:315]: The original file was a string with placeholders that I haven't implemented. I will do that now.

File: wdl2cwl/main.py (hunk 11 above)
Comment (Th3nn3ss) [reply wdl2cwl/main.py:66]: Yeah. I saw this and the required_inputs to and it looks interesting. Maybe we can explore it with another PR. Although I observed that the available_inputs don't carry the information of the namespace they are from. Which seems important for determining the ID of the input. So in the case where the import is more than one, we can't effectively return the right namespace for the right available_input.

I will look out for this when exploring files with more than one imports and see how the available inputs can be used.

Comment (mr-c) [reply wdl2cwl/main.py:66]: I think ` required_inputs ` is more for those running a WDL workflow, so we shouldn't look at it.

Comment (mr-c) [reply wdl2cwl/main.py:66]: > Maybe we can explore it with another PR.

Can you make an issue then?

Do that and resolve @kinow 's comment and I'll merge this, thanks!

**Pull request merged**